process/notifier/sovereignNotifier_test.go

```diff
@@ -0,0 +1,342 @@
+					},
+				},
+			},
+		}
+
+		err := sn.Notify(outportBlock)
+		require.NotNil(t, err)
+		require.True(t, strings.Contains(err.Error(), invalidHash))
+	})
+
+	t.Run("notify invalid header type", func(t *testing.T) {
+		args := createArgs()
+		sn, _ := NewSovereignNotifier(args)
+
+		blockData := createBlockData(args.Marshaller)
+		blockData.HeaderType = string(core.ShardHeaderV1)
+		outportBlock := &outport.OutportBlock{
+			BlockData:       blockData,
+			TransactionPool: &outport.TransactionPool{},
+		}
+
+		err := sn.Notify(outportBlock)
+		require.NotNil(t, err)
+		require.True(t, strings.Contains(err.Error(), errInvalidHeaderTypeReceived.Error()))
+		require.True(t, strings.Contains(err.Error(), blockData.HeaderType))
+	})
+
+	t.Run("notify invalid header bytes", func(t *testing.T) {
+		args := createArgs()
+		sn, _ := NewSovereignNotifier(args)
+
+		blockData := createBlockData(args.Marshaller)
+		blockData.HeaderBytes = []byte("invalid bytes")
+
+		outportBlock := &outport.OutportBlock{
+			BlockData:       blockData,
+			TransactionPool: &outport.TransactionPool{},
+		}
+
+		err := sn.Notify(outportBlock)
+		require.NotNil(t, err)
+	})
+}
+
+func TestSovereignNotifier_ConcurrentOperations(t *testing.T) {
+	t.Parallel()
+
+	addr1 := []byte("addr1")
+	txHash1 := []byte("hash1")
+
+	headerV2 := &block.HeaderV2{
+		Header:            &block.Header{},
+		ScheduledRootHash: []byte("root hash"),
+	}
+	extendedShardHeader := &block.ShardHeaderExtended{
+		Header: headerV2,
+		IncomingMiniBlocks: []*block.MiniBlock{
+			{
+				TxHashes:        [][]byte{txHash1},
+				ReceiverShardID: core.SovereignChainShardId,
+				SenderShardID:   0,
+				Type:            block.TxBlock,
+				Reserved:        nil,
+			},
+		},
+	}
+
+	args := createArgs()
+	args.SubscribedAddresses = [][]byte{addr1}
+
+	sn, _ := NewSovereignNotifier(args)
+
+	headerBytes, err := args.Marshaller.Marshal(headerV2)
+	require.Nil(t, err)
+	outportBlock := &outport.OutportBlock{
+		BlockData: &outport.BlockData{
+			HeaderBytes: headerBytes,
+			HeaderType:  string(core.ShardHeaderV2),
+		},
+		TransactionPool: &outport.TransactionPool{
+			Transactions: map[string]*outport.TxInfo{
+				hex.EncodeToString(txHash1): {
+					Transaction: &transaction.Transaction{
+						RcvAddr: addr1,
+					},
+					ExecutionOrder: 0,
+				},
+			},
+		},
+	}
+
+	n := 100
+	for i := 0; i < n; i++ {
+		switch i % 2 {
+		case 0:
+			go func() {
+				errNotify := sn.Notify(outportBlock)
+				require.Nil(t, errNotify)
+			}()
+		case 1:
+			go func() {
+				handler := &testscommon.ExtendedHeaderHandlerStub{
+					ReceivedExtendedHeaderCalled: func(header *block.ShardHeaderExtended) {
+						require.Equal(t, extendedShardHeader, header)
+					},
+				}
+
+				errRegister := sn.RegisterHandler(handler)
+				require.Nil(t, errRegister)
+			}()
+		default:
+			require.Fail(t, "should not have entered here")
+		}
+	}
+
+	time.Sleep(time.Millisecond * 200)
```

Comment:
the sync.WaitGroup approach would be better here.. but this seems to be fine for this test

Reply:
Definitely better with waitGroup. Applied your suggestion